File: indexer/extract/generic/testdata/typescript/sample.ts
import { Injectable } from "@angular/core";
import type { Repo } from "./repo";
import * as path from "path";
export { Service as Svc } from "./service";

/** Storage abstraction. */
export interface Store<T> extends Reader {
  get(id: string): T;
}

export type Id = string | number;

export enum Color {
  Red,
  Green = "green",
}

@Injectable()
export class UserService extends BaseService implements Store<User>, Disposable {
  private readonly cache = new Map<string, User>();
  protected static instances: number = 0;

  constructor(private repo: Repo) {
    super();
  }

  public get(id: string): User {
    return this.repo.find(id);
  }

  async dispose(): Promise<void> {
    path.join("a");
    await flush();
  }
}

export abstract class Shape {
  abstract area(): number;
}

namespace Geometry {
  export function area(r: number): number {
    return Math.PI * r * r;
  }
}

declare module "legacy" {
  export function old(): void;
}

export const DEFAULT_ID: Id = 0;
let counter = 0;

function helper(x: number): number {
  return x + counter;
}

File: indexer/extract/generic/testdata/c/sample.c
#include <stdio.h>
#include "store.h"

#define MAX_ITEMS 10
#define SQUARE(x) ((x) * (x))

/* A stored item. */
struct item {
    int id;
    char *name;
};

typedef struct item item_t;

enum color { RED, GREEN };

static int counter = 0;
int total;

int store_add(struct item *it);

// Adds an item.
int store_add(struct item *it) {
    printf("%d\n", it->id);
    counter++;
    return helper(it->id);
}

static void helper_internal(void) {
    store_add(NULL);
}
